### PR TITLE
add 64-bit intrin

### PR DESCRIPTION
**Pull request by lionkunonly**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under OpenCV (BSD) License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ ] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by lionkunonly** [comment]
#306 

Sorry for the late PR. The 64-bit intrin is added. The result of experiments will be added in the following. The function `cv.test_hal_intrin_all()`  is executed successfully. If I miss some functions, please let me know.

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp [hunk 1]
```diff
@@ -2684,21 +2556,19 @@ inline _Tpvec func(const _Tpvec& a, const _Tpvec& b) \
...
+// #define OPENCV_HAL_IMPL_WASM_MINMAX_64f_FUNC(func) \
+// inline v_float64x2 func(const v_float64x2& a, const v_float64x2& b) \
+// { \
+//     fallback::v_float64x2 a_(a), b_(b); \
+//     return fallback::func(a_, b_); \
+// }
+
+// OPENCV_HAL_IMPL_WASM_MINMAX_64f_FUNC(v_min)
+// OPENCV_HAL_IMPL_WASM_MINMAX_64f_FUNC(v_max)
+// #endif
```
Comment (huningxin) [inline modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: Why did you add those commented lines?

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp [hunk 2]
```diff
@@ -3318,30 +3131,22 @@ OPENCV_HAL_IMPL_WASM_CHECK_SIGNS(v_int16x8, i16x8, short)
...
+//     return fallback::v_signmask(a_);
+// }
+// inline bool v_check_all(const v_float64x2& a)
+// {
+//     return wasm_i8x16_all_true((__i64x2)(a.val) < (__i64x2)(wasm_i64x2_splat(0)));
+// }
+// inline bool v_check_any(const v_float64x2& a)
+// {
+//     return wasm_i8x16_any_true((__i64x2)(a.val) < (__i64x2)(wasm_i64x2_splat(0)));;
+// }
```
Comment (huningxin) [inline modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: ditto

**Review by huningxin — COMMENTED** [review]
Thanks for the PR. It looks good with some nits. As a summary, could you please list what 64 bit intrinsics are supported now? Are there any missing ones? What's the overall status?

**Comment by lionkunonly** [comment]
64 bit intrinsics supported now:
```
For Float 64:
v_float64x2
v_absdiff()
v_lut()
v_setzero_f64 ()
v_reinterpret_as_f64 ()
v_setall_f64()
 v_setzero_f64()

For Int 64
v_int64x2
v_setzero_s64()
v_setall_s64()
v_reinterpret_as_s64()
v_dotprod_expand
v_shl()
v_shr()
v_pack()
v_rshr_pack()
v_pack_store()

For Uint 64:
v_uint64x2
v_setzero_u64()
v_setall_u64()
v_reinterpret_as_u64()
v_shl()
v_shr()
v_dotprod_expand()
v_pack()
v_rshr_pack()
v_pack_store()
```

In my opinions, some functions can be implemented because of the 64-bit data is supported, like the `v_setall_s64()`. Some functions can be implemented for the new function. For example, the function `v_shl()` is implemented in wasm because the function `wasm_i64x2_shl()` is supported. 

There are some 64-bit intrinsics can not be implemented, like `v_cvt_f64()`, `v_cvt_f64_high` , `v_reduce_sum`

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp (hunk 1 above)
Comment (lionkunonly) [reply modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: They are deleted

File: modules/core/include/opencv2/core/hal/intrin_wasm.hpp (hunk 2 above)
Comment (lionkunonly) [reply modules/core/include/opencv2/core/hal/intrin_wasm.hpp]: They are deleted

**Review by terfendail — APPROVED** [review]

**Comment by huningxin** [comment]
> There are some 64-bit intrinsics can not be implemented, like `v_cvt_f64()`, `v_cvt_f64_high` , `v_reduce_sum`

@lionkunonly , could you please list those 64-bt intrinsics that are not implemented? And the corresponding missing feature of WASM SIMD?

**Comment by lionkunonly** [comment]
The table that 64-bit intrinsics that are not implemented:
|   Intrinsics |  Missing Features  |
|---|-----|
| CMP_OP  `==`,`!=`, `<`, `>`, `<=`, `>=` for int64. uint64 | missing function `wasm_i64x2_le`, `wasm_i64x2_le` and so on |
| `v_magnitude`, `v_sqr_magnitude` | missing squrare function function. Only function `wasm_f32x4_sqrt`, `wasm_f64x2_sqrt` exist |
| `v_store_low`, `v_store_high` | function `wasm_v128_store` does not support half of the data |
| `v_load_halves` | missing function to load data from two memory blocks |
| `v_signmask` | missing corresponding function |
| `v_check_all`, `v_check_any` for int64 and uint64 | missing functions to convert 64-bit to 32-bit |
| `v_cvt_f64`, `v_cvt_f64_high`, `v_cvt_f64` | missing function  `wasm_f64x2_convert_i64x2`, `v128_cvti32x4_i64x2` |

**Comment by terfendail** [comment]
@lionkunonly Could you please try to implement `v_magnitude` `v_sqr_magnitude` via SIMD multiplication and `v_check_all/any` via masked u32/i32 implementation?

**Comment by lionkunonly** [comment]
>Could you please try to implement `v_magnitude` `v_sqr_magnitude` via SIMD multiplication 


@terfendail  Thanks for your suggestions. I have implemented the `v_magnitude` `v_sqr_magnitude` via SIMD multiplication and SIMD squareRoot function. I also implement the `v_muladd`, too. Please check it, thanks.

>`v_check_all/any` via masked u32/i32 implementation?


I do not understand what the `masked u32/i32 implementation` means. Could you please give me more details? The implementation of `v_check_all/any` for u32/i32 uses the function `wasm_i32x4_lt`. And the function `wasm_i64x2_lt` does not exist.

**Comment by terfendail** [comment]
We could reinterpret i64x2 as i32x4. For negative i64 value higher i32 part will also be negative while lower part could be masked to 0x0 or 0xffffffff.

**Comment by lionkunonly** [comment]
@terfendail  I get your idea and I have implemented it. But I want to verify the `v_check_all` and `v_check_any`. Cloud you introduce some APIs that uses the two intrinsic?

**Comment by lionkunonly** [comment]
The `v_check_any` and `v_check_all` are updated, please check them.

**Comment by terfendail** [comment]
'v_check_any/all' intrinsics are used in 'Canny' and 'findContour' functions but they aren't highly used so the performance impact could be quite small.

**Review by terfendail — APPROVED** [review]
Looks fine for me. Thanks

**Review by huningxin — APPROVED** [review]
Looks good. Thanks!
